### PR TITLE
RPC Client trait and feature reorganization

### PR DESCRIPTION
**Pull request by trevor-crypto**
Apologies for such a large pull request. Please let me know if any of these changes are undesirable or need a major change. Thanks

Goals:
- Create an `RpcClient` trait that can be implemented externally to support any RPC backend (http). 
- Add support for ECDSA keys

Overview of changes:
- `RpcClient` trait 
- `Subscriber` trait, for future support of a client with subscriber/event backend (alternate websocket client maybe?)
- Move websocket client into `WsRpcClient` and implement `RpcClient` trait for it
- `ws-client` feature which enables `WsRpcClient` and uses `ws` crate. Enabled by default.
- reorganize modules to separate `std` feature better
- Modify extrinsic macro to support ECDSA keys 

Creating an `Api` is a breaking change, since now you need to specify a client. ex. `Api::<(), _>::new(client)`.

### CODE REVIEW

**Comment by clangenb** [comment]
On first sight, this looks how I imagined it 👍. I will make a thorough review soon.

**Comment by trevor-crypto** [comment]
Going to try and get the build checks / examples to pass. Might need some assistance if I can't get it :flushed:

**Comment by clangenb** [comment]
> Going to try and get the build checks / examples to pass. Might need some assistance if I can't get it 😳

This could be a dependency issue. https://github.com/scs/substrate-api-client-test-node might not yet be compatible with the substrate version that you reference in Cargo.lock. Hmm, annoying to have this dependency in the examples...

I will update the node as soon as possible, but I have some urgent stuff to do regarding our parachain launch this week.

**Comment by clangenb** [comment]
Thanks for fixing the tests!

File: src/extrinsic/mod.rs [hunk 1]
```diff
@@ -88,12 +91,11 @@ macro_rules! compose_extrinsic_offline {
 
         let signature = raw_payload.using_encoded(|payload| $signer.sign(payload));
 
-        let mut arr: [u8; 32] = Default::default();
-        arr.clone_from_slice($signer.public().as_ref());
+        let multi_signer: MultiSigner = $signer.public().into();
```
Comment (clangenb) [inline src/extrinsic/mod.rs:94]: Very nice, I hated this `arr.clone_from_slice`

File: src/std/rpc/json_req.rs [hunk 2]
```diff
@@ -107,10 +107,18 @@ pub fn state_get_keys(key: StorageKey, at_block: Option<Hash>) -> Value {
...
 
+pub fn author_submit_extrinsic(xthex_prefixed: &str) -> Value {
+    author_submit_extrinsic_with_id(xthex_prefixed, REQUEST_TRANSFER)
+}
+
 pub fn author_submit_and_watch_extrinsic(xthex_prefixed: &str) -> Value {
     author_submit_and_watch_extrinsic_with_id(xthex_prefixed, REQUEST_TRANSFER)
 }
 
+pub fn author_submit_extrinsic_with_id(xthex_prefixed: &str, id: u32) -> Value {
```
Comment (clangenb) [inline src/std/rpc/json_req.rs:118]: I don't get the ids, do you mind explaining it to me? :)

**Review by clangenb — APPROVED** [review]

**Comment by clangenb** [comment]
Very nice. I think this is ready to be merged. Also thanks for the cleanup, although the file reorganizations rendered the review process a bit hard.

* Only one remark: I think it might be nice to have a `Api::default_with_url(url: )` constructor, which instantiates the `WsRpcClient` under the hood, but this is only a detail not worth preventing the merge.

File: src/extrinsic/mod.rs (hunk 1 above)
Comment (trevor-crypto) [reply src/extrinsic/mod.rs:94]: Allows for ECDSA signing too!

File: src/std/rpc/json_req.rs (hunk 2 above)
Comment (trevor-crypto) [reply src/std/rpc/json_req.rs:118]: I think it's just a Json RPC protocol thing. I copy pasted from what was already there :blush:

**Comment by trevor-crypto** [comment]
@clangenb thanks a lot for taking the time to review! I pushed another commit with the default constructor idea.

Sorry for the major refactoring...it may be easier to see the structural changes if you pull the branch and poke around. I did not change any import logic other than the `MultiSigner` changes.

**Comment by brenzi** [comment]
I'd like to see our local CI test against a testnode succeed before merging. please be patient, need to fix CI

File: src/std/rpc/json_req.rs (hunk 2 above)
Comment (clangenb) [reply src/std/rpc/json_req.rs:118]: Haha, same with me. :P

**Comment by brenzi** [comment]
@trevor-crypto thanks for your contribution. we're almost ready to merge. can you please run `cargo fmt` and commit? The missing fmt is the reason for jenkins to fail currently

**Comment by clangenb** [comment]
@trevor-crypto Thank you very much for this PR! Sorry that it took so long until it was merged.

**Comment by trevor-crypto** [comment]
Thanks everyone for the help!

**Pull request merged**